community/detectors/rce/apache_tomcat_cve_2017_12617/src/main/java/com/google/tsunami/plugins/detectors/rce/cve201712617/Cve201712617Detector.java

```diff
@@ -0,0 +1,204 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.flogger.GoogleLogger;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.util.Timestamps;
+import com.google.tsunami.common.data.NetworkServiceUtils;
+import com.google.tsunami.common.net.http.HttpClient;
+import com.google.tsunami.common.net.http.HttpHeaders;
+import com.google.tsunami.common.net.http.HttpResponse;
+import com.google.tsunami.common.time.UtcClock;
+import com.google.tsunami.plugin.PluginType;
+import com.google.tsunami.plugin.VulnDetector;
+import com.google.tsunami.plugin.annotations.PluginInfo;
+import com.google.tsunami.plugin.payload.Payload;
+import com.google.tsunami.plugin.payload.PayloadGenerator;
+import com.google.tsunami.proto.DetectionReport;
+import com.google.tsunami.proto.DetectionReportList;
+import com.google.tsunami.proto.DetectionStatus;
+import com.google.tsunami.proto.NetworkService;
+import com.google.tsunami.proto.PayloadGeneratorConfig;
+import com.google.tsunami.proto.Severity;
+import com.google.tsunami.proto.TargetInfo;
+import com.google.tsunami.proto.Vulnerability;
+import com.google.tsunami.proto.VulnerabilityId;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.UUID;
+import javax.inject.Inject;
+
+/** A {@link VulnDetector} that detects highly critical RCE CVE-2017-12617. */
+@PluginInfo(
+    type = PluginType.VULN_DETECTION,
+    name = "CVE-2017-12617 Detector",
+    version = "0.1",
+    description = "Detects CVE-2017-12617 RCE vulnerability in Apache Tomcat.",
+    author = "Leonardo Tamiano (leonardo.tamiano@mindedsecurity.com)",
+    bootstrapModule = Cve201712617DetectorBootstrapModule.class)
+public final class Cve201712617Detector implements VulnDetector {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting static final String VULNERABILITY_REPORT_PUBLISHER = "TSUNAMI_COMMUNITY";
+  @VisibleForTesting static final String VULNERABILITY_REPORT_ID = "CVE_2017-12617";
+
+  @VisibleForTesting
+  static final String VULNERABILITY_REPORT_TITLE = "Apache Tomcat RCE CVE-2017-12617";
+
+  @VisibleForTesting
+  static final String VULN_DESCRIPTION =
+      "Tomcat instances containing a servlet context configured with readonly=false within the"
+          + " web.xml configuration, allow unauthenticated actors to upload arbitrary JSP files to"
+          + " the server via a specially crafted request. The uploaded JSP file could then be"
+          + " requested and any code it contained would be executed by the server, leading to"
+          + " Remote Code Execution (RCE).";
+
+  @VisibleForTesting
+  static final String RECOMMENDATION =
+      "Update vulnerable instances and ensure that readonly is set to true for the default servlet,"
+          + " and for the webdav servlet. Block HTTP Methods that allow untrusted users to modify"
+          + " server's resources";
```

Comment:
Please add upgrade to non-vulnerable versions of Apache Tomcat (versions.....) as part of recommendations as well.

Reply:
Added.